File: flutter/packages/hatake_yaml/test/vocabularies_test.dart
import 'package:hatake_core/hatake_core.dart';
import 'package:hatake_yaml/hatake_yaml.dart';
import 'package:test/test.dart';

/// `app.vocabularies` に書いた語彙が、**使う所へ実体で入る**こと。
///
/// 見本2本で、同じコード表を 3回・2回・3回 書いていた。片方だけ直すと画面ごとに
/// 違う字が出て、しかも直すまで誰も気づかない。だから1か所に書いて名前で指す。
///
/// 展開は**読み込み時**なので、この先（Renderer・CSV・紙）は語彙を知らない。
/// ここで実体が入っていることを確かめれば、下流は今までどおりで正しく出る。
const _yaml = '''
dsl_version: "1.0"
app:
  id: orders
  title: 受注
  home: orders
  vocabularies:
    - name: orderStatus
      options:
        - { value: draft, label: 作成中 }
        - { value: shipped, label: 出荷済 }
  menu:
    - { id: orders, label: 受注一覧, page: orders }
  pages:
    - type: crud
      id: orders
      title: 受注一覧
      repository: orderRepository
      key: orderNo
      search:
        filters:
          - { field: status, label: 受注状態, optionsOf: orderStatus }
      table:
        columns:
          - { field: orderNo, label: 受注番号 }
          - { field: status, label: 受注状態, optionsOf: orderStatus }
      form:
        sections:
          - fields:
              - { field: orderNo, label: 受注番号, type: text, required: true }
              - { field: status, label: 受注状態, type: select, optionsOf: orderStatus }
''';

void main() {
  test('列・検索条件・入力項目のどれにも、実体の選択肢が入る', () {
    final app = parseAppYaml(_yaml);
    final page = app.pages.first as CrudPageDefinition;

    final column = page.table.columns.firstWhere((c) => c.field == 'status');
    expect(column.options.map((o) => o.label), ['作成中', '出荷済']);

    final filter = page.search!.filters.first;
    expect(filter.options.map((o) => o.label), ['作成中', '出荷済']);

    final field = page.form.sections.first.fields
        .firstWhere((f) => f.field == 'status');
    expect(field.options.map((o) => o.label), ['作成中', '出荷済']);
  });

  test('列が名指しした語彙で、そのまま字になる', () {
    final app = parseAppYaml(_yaml);
    final page = app.pages.first as CrudPageDefinition;
    final column = page.table.columns.firstWhere((c) => c.field == 'status');

    // 借りる仕掛けを通さなくても、自分の持ち物で出る。
    expect(cellText(FormatterRegistry(), const [], column, 'shipped'), '出荷済');
  });

  test('その場に書いた options が勝つ', () {
    final app = parseAppYaml(_yaml.replaceFirst(
      '- { field: status, label: 受注状態, optionsOf: orderStatus }\n      form:',
      '- { field: status, label: 受注状態, optionsOf: orderStatus, '
          'options: [{ value: shipped, label: 発送済 }] }\n      form:',
    ));
    final page = app.pages.first as CrudPageDefinition;
    final column = page.table.columns.firstWhere((c) => c.field == 'status');
    expect(column.options.single.label, '発送済');
  });

  test('引けない名前のときは、空の並びを置かない', () {
    // 空を置くと「書いたのに効かない」が検証にも出なくなる。
    final app = parseAppYaml(_yaml.replaceAll('optionsOf: orderStatus', 'optionsOf: nope'));
    final page = app.pages.first as CrudPageDefinition;
    expect(page.table.columns.firstWhere((c) => c.field == 'status').options, isEmpty);
  });

  test('語彙を1つも書いていない定義は、今までどおり読める', () {
    final app = parseAppYaml(_yaml
        .replaceAll(RegExp(r'  vocabularies:[\s\S]*?  menu:'), '  menu:')
        .replaceAll(', optionsOf: orderStatus', ''));
    expect(app.pages, hasLength(1));
  });
}

File: typescript/test/vocabularies.test.ts
import { describe, expect, it } from "vitest";

import { expandVocabularies, vocabulariesOf } from "../src/vocabularies.js";
import { findVocabularyProblems } from "../src/vocabularyChecks.js";

const VOC = [
  {
    name: "orderStatus",
    options: [
      { value: "draft", label: "作成中" },
      { value: "shipped", label: "出荷済" },
    ],
  },
];

const app = (page: unknown, vocabularies: unknown[] = VOC) => ({
  app: { id: "orders", title: "受注", vocabularies, pages: [page] },
});

describe("語彙の展開", () => {
  it("列の optionsOf が、実体の並びに置き換わる", () => {
    const out = expandVocabularies(
      app({ table: { columns: [{ field: "status", optionsOf: "orderStatus" }] } }),
    ) as any;
    const column = out.app.pages[0].table.columns[0];
    expect(column.options).toEqual(VOC[0].options);
    // 名前は残す（どこから来たのかが読めなくなると、直す所が分からない）。
    expect(column.optionsOf).toBe("orderStatus");
  });

  it("入力項目と検索条件でも同じように置き換わる", () => {
    const out = expandVocabularies(
      app({
        form: { sections: [{ fields: [{ field: "status", optionsOf: "orderStatus" }] }] },
        search: { filters: [{ field: "status", optionsOf: "orderStatus" }] },
      }),
    ) as any;
    expect(out.app.pages[0].form.sections[0].fields[0].options).toHaveLength(2);
    expect(out.app.pages[0].search.filters[0].options).toHaveLength(2);
  });

  it("その場に書いた options が勝つ（両方書いてあるとき）", () => {
    const out = expandVocabularies(
      app({
        table: {
          columns: [
            {
              field: "status",
              optionsOf: "orderStatus",
              options: [{ value: "draft", label: "下書き" }],
            },
          ],
        },
      }),
    ) as any;
    expect(out.app.pages[0].table.columns[0].options).toEqual([
      { value: "draft", label: "下書き" },
    ]);
  });

  it("引けない名前は、空の並びを置かずにそのまま残す", () => {
    // 空を置くと「書いたのに効かない」が検証にも出なくなる。
    const out = expandVocabularies(
      app({ table: { columns: [{ field: "status", optionsOf: "nope" }] } }),
    ) as any;
    expect(out.app.pages[0].table.columns[0].options).toBeUndefined();
  });

  it("元の定義は変えない", () => {
    const before = app({
      table: { columns: [{ field: "status", optionsOf: "orderStatus" }] },
    });
    expandVocabularies(before);
    expect((before as any).app.pages[0].table.columns[0].options).toBeUndefined();
  });

  it("同じ名前が2回あれば、先に書いたほうが残る", () => {
    const found = vocabulariesOf(
      app({}, [
        VOC[0],
        { name: "orderStatus", options: [{ value: "shipped", label: "発送済" }] },
      ]),
    );
    expect(found.get("orderStatus")).toEqual(VOC[0].options);
  });

  it("語彙が1つも無ければ、何も触らない", () => {
    const before = { app: { id: "a", title: "b", pages: [] } };
    expect(expandVocabularies(before)).toBe(before);
  });
});

describe("語彙の食い違い", () => {
  it("指した先が無ければ言う。近い名前があれば添える", () => {
    const found = findVocabularyProblems(
      app({ table: { columns: [{ field: "status", optionsOf: "orderStatas" }] } }),
    );
    expect(found).toHaveLength(1);
    expect(found[0].rule).toBe("unknown-vocabulary");
    expect(found[0].near).toBe("orderStatus");
  });

  it("遠い名前は勧めない（関係の無い名前で迷わせない）", () => {
    const found = findVocabularyProblems(
      app({ table: { columns: [{ field: "status", optionsOf: "paymentTerms" }] } }),
    );
    expect(found[0].near).toBeUndefined();
    expect(found[0].known).toEqual(["orderStatus"]);
  });

  it("options と両方書いてあれば言う", () => {
    const found = findVocabularyProblems(
      app({
        table: {
          columns: [
            { field: "status", optionsOf: "orderStatus", options: [{ label: "x" }] },
          ],
        },
      }),
    );
    expect(found.map((one) => one.rule)).toEqual(["vocabulary-shadowed"]);
  });

  it("名前が重なっていれば言う", () => {
    const found = findVocabularyProblems(
      app({}, [VOC[0], { name: "orderStatus", options: [] }]),
    );
    expect(found.map((one) => one.rule)).toEqual(["duplicate-vocabulary"]);
  });

  it("語彙そのものの中は見ない（`options` は在って当たり前）", () => {
    expect(findVocabularyProblems(app({}))).toEqual([]);
  });
});
